File: src/boundary/eslint-scope/enter-catch.ts
import { DEFINITION_TYPE } from "../../analyzer/definition-type.js";
import type { AstNode } from "../../ir/primitive/ast-node.js";
import { AST_TYPE } from "../../parser/ast-type.js";
import type { DiagnosticCollector } from "../../util/diagnostic.js";
import { collectBindingIdentifiers } from "./declare/collect-binding-identifiers.js";
import { declareVariable } from "./declare/declare-variable.js";
import { hoistDeclarations } from "./hoisting/hoist-declarations.js";
import { isNodeLike } from "./is-node-like.js";
import type { ScopeManager } from "./manager.js";
import type { NodeLike } from "./node-like.js";
import type { AnalysisVisitor } from "./visitor.js";
import type { PathEntry } from "./walk/path-entry.js";

export function enterCatch(
  node: NodeLike,
  parent: NodeLike | null,
  key: string | null,
  path: readonly PathEntry[],
  manager: ScopeManager,
  raw: string,
  diagnostics: DiagnosticCollector,
  visitor: AnalysisVisitor,
): void {
  const scope = manager.push("catch", node as unknown as AstNode);
  visitor.onScope?.({ scope, parent, key, path });
  const param = node["param"];
  if (isNodeLike(param)) {
    const idents = collectBindingIdentifiers(param as unknown as AstNode);
    for (const ident of idents) {
      declareVariable(
        scope,
        ident,
        DEFINITION_TYPE.CatchClause,
        node as unknown as AstNode,
        null,
      );
    }
  }
  const body = node["body"];
  if (isNodeLike(body) && body.type === AST_TYPE.BlockStatement) {
    const stmts = body["body"];
    if (Array.isArray(stmts)) {
      hoistDeclarations(stmts, scope, raw, diagnostics);
    }
  }
}

File: crates/unsnarl-boundary-eslint-scope/src/enter_catch.rs
//! Push a `Catch` scope, declare the caught parameter, and hoist
//! the catch body's declarations.
//!
//! Mirrors `enterCatch` in
//! `ts/src/boundary/eslint-scope/enter-catch.ts`. The TS port reads
//! `node["param"]` and walks its bindings; the Rust port reads
//! `CatchClause.param: Option<CatchParameter>` and walks
//! `CatchParameter.pattern: BindingPattern<'_>`. The body of the
//! catch is `Box<BlockStatement>` in oxc (vs. plain `BlockStatement`
//! in TS), so we hoist `body.body`.

use oxc_ast::ast::CatchClause;

use unsnarl_ir::ids::ScopeId;
use unsnarl_ir::primitive::AstNode;
use unsnarl_ir::scope_type::ScopeType;
use unsnarl_ir::DefinitionType;
use unsnarl_oxc_parity::AstType;

use crate::declare::collect_binding_identifiers::collect_binding_identifiers;
use crate::hoisting::hoist_declarations::hoist_declarations;
use crate::state::{declare_variable, push_scope, ScopeBuilderState};

pub(crate) fn enter_catch(
    state: &mut ScopeBuilderState,
    catch: &CatchClause<'_>,
    raw: &str,
) -> ScopeId {
    let catch_node = AstNode {
        r#type: AstType::CatchClause,
        span: catch.span,
    };
    let scope_id = push_scope(state, ScopeType::Catch, catch_node.clone());
    if let Some(param) = catch.param.as_ref() {
        for ident in collect_binding_identifiers(&param.pattern) {
            declare_variable(
                state,
                scope_id,
                ident,
                DefinitionType::CatchClause,
                catch_node.clone(),
                None,
            );
        }
    }
    hoist_declarations(state, &catch.body.body, scope_id, raw);
    scope_id
}

#[cfg(test)]
#[path = "enter_catch_test.rs"]
mod enter_catch_test;
